composeApp/src/commonMain/kotlin/sp/bvantur/tasky/register/presentation/RegisterViewModel.kt

```diff
@@ -0,0 +1,120 @@
+import sp.bvantur.tasky.core.presentation.ViewModelViewStateHandlerImpl
+import kotlin.time.Duration.Companion.milliseconds
+
+typealias OnRegisterUserAction = (RegisterUserAction) -> Unit
+
+@OptIn(FlowPreview::class)
+class RegisterViewModel(
+    private val dispatcherProvider: DispatcherProvider,
+    private val validateNameUseCase: ValidateNameUseCase,
+    private val validateEmailUseCase: ValidateEmailUseCase,
+    private val validatePasswordUseCase: ValidatePasswordUseCase
+) : ViewModel(),
+    ViewModelUserActionHandler<RegisterUserAction>,
+    ViewModelViewStateHandler<RegisterViewState> by ViewModelViewStateHandlerImpl(
+        RegisterViewState(),
+        dispatcherProvider
+    ) {
+
+    private val textDebounceFlow = MutableSharedFlow<Pair<String, suspend (String) -> Unit>>()
+
+    init {
+        textDebounceFlow
+            .debounce(500.milliseconds)
+            .onEach { (text, action) ->
+                action(text)
+            }
+            .launchIn(CoroutineScope(dispatcherProvider.default))
+    }
+
+    override fun onUserAction(userAction: RegisterUserAction) {
+        when (userAction) {
+            is RegisterUserAction.NameChanged -> onNameChanged(userAction.value)
+            is RegisterUserAction.EmailChanged -> onEmailChanged(userAction.value)
+            is RegisterUserAction.PasswordChanged -> onPasswordChanged(userAction.value)
+            is RegisterUserAction.RegisterUser -> onRegisterUser(userAction.name, userAction.email, userAction.password)
+            RegisterUserAction.NavigateBack -> onNavigateBack()
+        }
+    }
+
+    private fun onNameChanged(value: String) {
+        viewModelScope.launch {
+            textDebounceFlow.emit(
+                value to {
+                    val isValid = validateNameUseCase.invoke(it)
+                    emitViewState(
+                        viewStateFlow.value.copy(isNameError = !isValid)
+                    )
+                }
+            )
+        }
+    }
+
+    private fun onEmailChanged(value: String) {
+        viewModelScope.launch {
+            textDebounceFlow.emit(
+                value to {
+                    val isValid = validateEmailUseCase.invoke(it)
+                    emitViewState(
+                        viewStateFlow.value.copy(isEmailError = !isValid)
+                    )
+                }
+            )
+        }
+    }
+
+    private fun onPasswordChanged(value: String) {
+        viewModelScope.launch {
+            textDebounceFlow.emit(
+                value to {
+                    val isValid = validatePasswordUseCase.invoke(it)
+                    emitViewState(
+                        viewStateFlow.value.copy(isPasswordError = !isValid)
+                    )
+                }
+            )
+        }
+    }
+
+    private fun onRegisterUser(name: String, email: String, password: String) {
+        viewModelScope.launch(dispatcherProvider.main) {
```

Comment:
You don’t need to explicitly pass in the main dispatcher since `viewModelScope` already runs on the main thread.

Reply:
It actually runs on `Dispatchers.Main.immediate` by default, so by passing in the `Main` dispatcher without immediate, you break the immediate dispatching behavior the ViewModel should usually have

Reply:
This is another habit that I have with other projects. Usually, I test ViewModels in my projects and use my dispatcher provider for every scope launch to prepare the code for testing. I will remove it from here and take that tip with me. Thank you! :)

Reply:
That's all good, but if it's just about replacing the main dispatcher for your tests (and not IO or Default), the `Dispatchers.setMain(testDispatcher)` already does that for you in your tests 👍

Reply:
Oh, I see...I didn't know that such a method exists for testing purposes. Thank you for that tip. I was usually exchanging my dispactherProvider with StandardTestDispatchers. I will check this approach and try to use it in my work projects. Thanks for that.